tests/MongoDB.Driver.Tests/Specifications/crud/CrudTestRunner.cs

```diff
@@ -19,55 +19,228 @@
+        private void DropCollection(string databaseName, string collectionName)
+        {
+            DriverTestConfiguration.Client.GetDatabase(databaseName).DropCollection(collectionName);
+        }
+
+        private void ExecuteOperation(IMongoClient client, IMongoDatabase database, IMongoCollection<BsonDocument> collection, BsonDocument operation, BsonDocument outcome, bool async)
         {
             var name = (string)operation["name"];
             var test = CrudOperationTestFactory.CreateTest(name);
+            bool isErrorExpected = operation.GetValue("error", false).AsBoolean;
 
             var arguments = (BsonDocument)operation.GetValue("arguments", new BsonDocument());
-            string reason;
-            if (!test.CanExecute(DriverTestConfiguration.Client.Cluster.Description, arguments, out reason))
+            test.SkipIfNotSupported(arguments);
+
+            test.Execute(client.Cluster.Description, database, collection, arguments, outcome, isErrorExpected, async);
+
+            if ((test.ActualError != null) != isErrorExpected)
             {
-                throw new SkipException(reason);
+                if (test.ActualError != null)
+                {
+                    throw test.ActualError;
+                }
+                else
+                {
+                    throw new Exception("The test was expected to throw an exception, but the executing was successful.");
+                }
             }
+        }
 
-            test.Execute(DriverTestConfiguration.Client.Cluster.Description, database, collection, arguments, outcome, async);
+        private IMongoCollection<BsonDocument> GetCollection(IMongoDatabase database, string collectionName, BsonDocument operation)
+        {
+            var collectionSettings = ParseCollectionOptions(operation);
+
+            var collection = database.GetCollection<BsonDocument>(
+                collectionName,
+                collectionSettings);
+
+            return collection;
+        }
+
+        private string GetCollectionName(BsonDocument operation)
+        {
+            if (operation.TryGetValue("collection_name", out var collectionName))
+            {
+                return collectionName.AsString;
+            }
+            else
+            {
+                return DriverTestConfiguration.CollectionNamespace.CollectionName;
+            }
+        }
+
+        private void ParseCollectionOptions(MongoCollectionSettings settings, BsonDocument collectionOptions)
+        {
+            foreach (var collectionOption in collectionOptions.Elements)
+            {
+                switch (collectionOption.Name)
+                {
+                    case "readConcern":
+                        settings.ReadConcern = ReadConcern.FromBsonDocument(collectionOption.Value.AsBsonDocument);
+                        break;
```

Comment:
Do we need to throw a `FormatException` if there are any options names we don't understand?

Reply:
Done